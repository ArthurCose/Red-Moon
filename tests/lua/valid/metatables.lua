local function implement_custom(table, name)
  table[name] = function(...) print(name, " called with args: ", ...) end
end

local m = {}
implement_custom(m, "__unm")
implement_custom(m, "__bnot")
implement_custom(m, "__add")
implement_custom(m, "__sub")
implement_custom(m, "__mul")
implement_custom(m, "__div")
implement_custom(m, "__idiv")
implement_custom(m, "__mod")
implement_custom(m, "__pow")
implement_custom(m, "__band")
implement_custom(m, "__bor")
implement_custom(m, "__bxor")
implement_custom(m, "__shl")
implement_custom(m, "__shr")
implement_custom(m, "__eq")
implement_custom(m, "__lt")
implement_custom(m, "__le")
implement_custom(m, "__concat")
implement_custom(m, "__len")
implement_custom(m, "__index")
implement_custom(m, "__newindex")
implement_custom(m, "__call")

local a = {}
setmetatable(a, m)

local noop = function(...) end

print("tables:")
noop(#a)
noop(-a)
noop(~a)
noop(a + 1, 1 + a)
noop(a - 1, 1 - a)
noop(a * 1, 1 * a)
noop(a / 1, 1 / a)
noop(a // 1, 1 // a)
noop(a % 1, 1 % a)
noop(a ^ 1, 1 ^ a)
noop(a & 1, 1 & a)
noop(a | 1, 1 | a)
noop(a ~ 1, 1 ~ a)
noop(a << 1, 1 << a)
noop(a >> 1, 1 >> a)
noop(a == 1, 1 == a)
noop(a ~= 1, 1 ~= a)
noop(a < 1, 1 < a)
noop(a > 1, 1 > a)
noop(a <= 1, 1 <= a)
noop(a >= 1, 1 >= a)
noop(a[1])
a[1] = 2
a(1, 2)

print("\n__index short circuit:")
local b = { value = 5 }
print(b.value)
setmetatable(b, m)
print(b.value)

print("\nsimple __index chain")
do
  local a, b, c = {}, {}, { value = 1 }
  setmetatable(a, { __index = b })
  setmetatable(b, { __index = c })
  print(a.value)
end
