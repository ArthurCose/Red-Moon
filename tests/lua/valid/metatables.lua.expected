tables:
__len	 called with args: 	table
__unm	 called with args: 	table
__bnot	 called with args: 	table
__add	 called with args: 	table	1
__add	 called with args: 	1	table
__sub	 called with args: 	table	1
__sub	 called with args: 	1	table
__mul	 called with args: 	table	1
__mul	 called with args: 	1	table
__div	 called with args: 	table	1
__div	 called with args: 	1	table
__idiv	 called with args: 	table	1
__idiv	 called with args: 	1	table
__mod	 called with args: 	table	1
__mod	 called with args: 	1	table
__pow	 called with args: 	table	1
__pow	 called with args: 	1	table
__band	 called with args: 	table	1
__band	 called with args: 	1	table
__bor	 called with args: 	table	1
__bor	 called with args: 	1	table
__bxor	 called with args: 	table	1
__bxor	 called with args: 	1	table
__shl	 called with args: 	table	1
__shl	 called with args: 	1	table
__shr	 called with args: 	table	1
__shr	 called with args: 	1	table
__eq	 called with args: 	table	1
__eq	 called with args: 	1	table
__eq	 called with args: 	table	1
__eq	 called with args: 	1	table
__lt	 called with args: 	table	1
__lt	 called with args: 	1	table
__lt	 called with args: 	1	table
__lt	 called with args: 	table	1
__le	 called with args: 	table	1
__le	 called with args: 	1	table
__le	 called with args: 	1	table
__le	 called with args: 	table	1
__index	 called with args: 	table	1
__newindex	 called with args: 	table	1	2
__call	 called with args: 	table	1	2

__index short circuit:
5
5

simple __index chain
1
